### ISSUE TITLE
[Bug] `GetInputTokens()` doesn't preserve the system prompt when we're over `max_window_size_`

### ISSUE THREAD

**Issue by hammer-ai** [mlc-ai/mlc-llm#836]
## 🐛 Bug

I believe I found a bug in the logic that gets triggered `GetInputTokens()` when we're over `max_window_size_`. This is discussed in detail here: https://discord.com/channels/1108078063931097149/1144632647918829699/1144762321558831254

The fix is in [this code](https://github.com/mlc-ai/mlc-llm/blob/main/cpp/llm_chat.cc#L447):
```
    // keep system
    if (this->conversation_.system.empty()) {
      all_prompt = GetConcatPrompt(prompts, 0, start_re_encode_pos);
    } else {
      all_prompt = GetConcatPrompt(prompts, 1, start_re_encode_pos);
    }
```
which should instead be:
```
    // keep system
    if (this->conversation_.system.empty()) {
      all_prompt = GetConcatPrompt(all_prompt, 0, start_re_encode_pos);
    } else {
      all_prompt = GetConcatPrompt(all_prompt, 1, start_re_encode_pos);
    }
```

This is because `prompts` only contains the messages that the user just sent (i.e. their last message to the LLM). Whereas `all_prompts` contains all messages of both the user and the LLM. When I make this change then I don't see the sudden coherence drop after this code is run, and can have long chats with characters using [https://www.hammerai.com/](https://www.hammerai.com).

## To Reproduce

1. Talk with the model for long enough to trigger the `max_window_size_` condition in `GetInputTokens()`
2. You will see the model forgets the system prompt and starts speaking nonsensical things or in a different language. In the example below it goes from being Thor to talking about Thailand:

![image](https://github.com/mlc-ai/mlc-llm/assets/143602265/695279c7-7cf8-4e46-b95f-85234a840b2b)

## Expected behavior

When we get past `max_window_size_` we correctly keep some of the old information.

## Environment

 - Platform (e.g. WebGPU/Vulkan/IOS/Android/CUDA): Metal
 - Operating system (e.g. Ubuntu/Windows/MacOS/...): MacOS
 - Device (e.g. iPhone 12 Pro, PC+RTX 3090, ...): MacOS
 - How you installed MLC-LLM (`conda`, source): Source
 - How you installed TVM-Unity (`pip`, source): `pip`
 - Python version (e.g. 3.10): 3.11.5
 - GPU driver version (if applicable):
 - CUDA/cuDNN version (if applicable):
 - TVM Unity Hash Tag (`python -c "import tvm; print('\n'.join(f'{k}: {v}' for k, v in tvm.support.libinfo().items()))"`, applicable if you compile models):
 - Any other relevant information:

## Additional context

Thank you!

**Comment by MasterJH5574** [comment]
Hi @hammer-ai, this was fixed by #828 :-)

**Comment by hammer-ai** [comment]
Awesome thank you!
